### PR TITLE
ENH: Allow arbitrary dtypes in binomial

### LINKED ISSUE #2011: np.random.binomial() fails with int64 'n' argument on x86/32-bit (Trac #1413)
_Original ticket http://projects.scipy.org/numpy/ticket/1413 on 2010-02-26 by @dwf, assigned to unknown._

Reported to the list by James Bergstra, confirmed by me:

```
>>> numpy.random.binomial(n=numpy.asarray([2,3,4], dtype='int64'),
p=numpy.asarray([.1, .2, .3], dtype='float64'))
```

produces:

```
TypeError: array cannot be safely cast to required type
```

It seems to be not only 32-bit specific but x86-specific. On a ppc  
machine, 32-bit mode, it behaves as expected:

```
dwf@morrislab:~$ python-32
Python 2.6.4 (r264:75706, Feb 16 2010, 21:03:46)
[GCC 4.0.1 (Apple Inc. build 5493)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
import numpy
numpy.__version__
'1.3.0'
numpy.random.binomial(n=numpy.asarray([2,3,4], dtype='int64'),  
p=numpy.asarray([.1,.2,.3], dtype='float64'))
array([1, 1, 2])
```

So it smells a bit like an endianness bug/problem with the definition of `NPY_LONG`.

I can confirm the bug on OS X/Intel 32-bit, and Linux x86-32 (both  
1.3.0 and most recent svn trunk), as well as its absence on Linux  
x86-64. The problem seems to be with this line in [source:trunk/numpy/random/mtrand/mtrand.pyx mtrand.pyx], line  
3306 in the trunk:

```
on = <ndarray>PyArray_FROM_OTF(n, NPY_LONG, NPY_ALIGNED)
```

### PR DESCRIPTION
**Pull request by bashtage**
Allow arbitrary dtypes as long as the values are in range
for long

closes #2011

### CODE REVIEW

File: numpy/random/mtrand/mtrand.pyx [hunk 1]
```diff
@@ -3698,7 +3698,8 @@ cdef class RandomState:
         ----------
         n : int or array_like of ints
             Parameter of the distribution, >= 0. Floats are also accepted,
-            but they will be truncated to integers.
+            but they will be truncated to integers. All values must be less
+            than or equal to np.iinfo(np.long).max
```
Comment (eric-wieser) [inline numpy/random/mtrand/mtrand.pyx]: `np.long` is not the same as NPY_LONG. the former is simply the built-in `long`, or `int` in python 3

**Comment by bashtage** [comment]
I figured this out in the code part, but what is the right way to
communicate the bound to the end user? Iinfo(int).max?

**Comment by eric-wieser** [comment]
Possibly `iinfo('l')`? Worth checking if that's correct.

Also, I suspect the tests will fall on win32, where long is only an int32 (or at least, `builtins.int` is)

**Comment by bashtage** [comment]
Using 'np.iinfo('l').max` seems to work across all all platforms.  It is also used for another constant for the maximum Poisson intensity, which indicates it is likely correct.

**Comment by eric-wieser** [comment]
I'd argue that this fix is not desirable, as it changes a dtype-based error (easy to catch with tests) with a value-based one.

**Comment by bashtage** [comment]
The function is a bit uneven in its treatment of inputs.

```python
np.random.binomial([3.0,10.2],0.5)
Out[26]: array([2, 7])

np.random.binomial(np.array([3.0,10.2]),0.2)
TypeError: Cannot cast array data from dtype('float64') to dtype('int32') according to the rule 'safe'

np.random.binomial(np.array([3.0,10.2],dtype=np.int32),0.2)
Out[28]: array([1, 6])

np.random.binomial(np.array([3.0,10.2],dtype=np.int64),0.2)
Cannot cast array data from dtype('int64') to dtype('int32') according to the rule 'safe'
```

File: numpy/random/mtrand/mtrand.pyx [hunk 2]
```diff
@@ -3771,16 +3772,21 @@ cdef class RandomState:
         cdef ndarray on, op
         cdef long ln
         cdef double fp
+        
+        on = <ndarray>np.array(n)
+        if np.any(np.greater(n, np.iinfo('l').max)):
```
Comment (charris) [inline numpy/random/mtrand/mtrand.pyx]: So this converts `n` to a numpy array, which is what `on` already is.

File: numpy/random/mtrand/mtrand.pyx [hunk 3]
```diff
@@ -3771,16 +3772,21 @@ cdef class RandomState:
...
+            raise OverflowError('n is out of range for C long '
+                                '(max : {0})'.format(np.iinfo('l').max))
+        if np.any(np.less(n, 0)):
```
Comment (charris) [inline numpy/random/mtrand/mtrand.pyx]: See above.

**Comment by charris** [comment]
I'm in general agreement with @eric-wieser here, and would actually prefer that the caller deal with the values before passing them. @rkern has also suggested making a pass through all of np.random at sometime in the future and dealing systematically with the long parameters.

**Comment by bashtage** [comment]
This is probably the wrong way -- converting to `int64` -- especially the ability to accept `int64` -- is a much more sensible approach IMO.

**Pull request closed without merging**